torch/utils/mem_tracker.py

```diff
@@ -0,0 +1,510 @@
+        reftype: RefType,
+        WINFO: WeakIdKeyDictionary,
+        update_existing: bool = False,
+    ) -> Self:
+        st = cls.get_untyped_storage(t)
+        if winfo := cls.get_winfo_if_exists(st, WINFO):
+            winfo.reftype = reftype
+            return winfo
+        elif update_existing:
+            raise ValueError("No existing winfo found")
+        else:
+            winfo = cls.create_winfo(st, reftype, WINFO)
+            return winfo
+
+
+def _get_mem_divisor(units: str) -> int:
+    return {"B": 1, "KB": 2**10, "MB": 2**20, "GB": 2**30}.get(units, 1)
+
+
+def _rounding_fn(value: int, divisor: int, precision: int) -> Union[float, int]:
+    return value if divisor == 1 else round(value / divisor, precision)
+
+
+def _print_snapshot(snapshot: Dict[torch.device, Dict[str, int]], units: str):
+    divisor = _get_mem_divisor(units)
+    for dev, dev_snap in snapshot.items():
+        print(
+            f"Device: {dev}",
+            *(
+                f"\t{k}: {_rounding_fn(v, divisor, 2)} {units}"
+                for k, v in dev_snap.items()
+            ),
+            sep="\n",
+        )
+
+
+class MemTracker(TorchDispatchMode):
+    """
+    A TorchDispatchMode that tracks the memory usage of parameters, buffers, activations, gradients, and optimizer states
+    within its context. It captures "memory snapshots" of the module's memory usage at various points in time
+    [pre_fw, post_fw, pre_bw, post_bw, peak_fw, peak_bw].
+
+    Attributes:
+        memory_tracking: A weakref key dictionary to store the memory statistics of each module. Each key
+        is a weak reference to a module, and each value is a `ModMemStats` object that stores the memory
+        statistics of the module.
+
+    Note:
+        The MemTracker should be used as a context manager. The modules, optimizers, and any other tensors created within
+        the context of MemTracker will be tracked by default. All the modules, optimizers, and tensors that need to be
+        tracked but are created outside the MemTracker should be registered using the `track_external` method. The
+        `track_external` method should be called before the MemTracker is used. Any tensors created outside the MemTracker
+        and not supplied to the `track_external` method will be tracked only after their first use inside the MemTracker.
+
+    Example usage:
+
+        .. code-block:: python
+
+            module = ...
+            optimizer = ...
+            inp = ...
+            mem_tracker = MemTracker()
+            mem_tracker.track_external(module, optimizer, inp)
+            with mem_tracker as mt:
+                optimizer.zero_grad()
+                loss = module(inp)
+                print("After Forward:")
+                mt.print_mem_snapshot()
+                loss.backward()
+                optimizer.step()
+            mt.print_mem_snapshot(peak = True)
+            mt.display_module_mem_stats(depth = 3, units = "MB")
```

Comment:
Given that this is only for "debugging" I would expect that we only need one that prints the summary.

Reply:
I have kept the two, because even for debugging, one might just want to look at the current snapshot or the peak memory snapshot not the entire module statistics.